### PR TITLE
[CLI] Only use referenced dependencies in the linkage table

### PR DESCRIPTION
**Pull request by stefan-mysten**
## Description 

When a package is added as a dependency in the Move.toml manifest file, it will become part of the linkage table regardless if any code in that package is actually referenced in the code or not. 

This PR fixes that by making sure that a package's linkage table contains only the dependencies that are actually referenced in the code by computing the list of dependencies from the modules' dependencies after all modules and their transitive dependencies are compiled rather than the manifest files.

## Test plan 

Note that now source validation will not allow to publish such a package (see below) because it cannot find the dependency on-chain.

Structure:
A_v1 -> published
A_v2 -> upgrade A_v1
B -> dependency on A_v2, but no code is called from A_v2. 

Locally: After the change
```
{
  "data": {
    "package": {
      "modules": {
        "nodes": [
          {
            "name": "b",
            "package": {
              "address": "0xb185657f36ad8af5c84e95884f674747bcfffe5bc51b98053bfbf2c349ef0633",
              "linkage": [
                {
                  "originalId": "0x0000000000000000000000000000000000000000000000000000000000000001",
                  "upgradedId": "0x0000000000000000000000000000000000000000000000000000000000000001",
                  "version": 1
                },
                {
                  "originalId": "0x0000000000000000000000000000000000000000000000000000000000000002",
                  "upgradedId": "0x0000000000000000000000000000000000000000000000000000000000000002",
                  "version": 1
                }
              ]
            }
          }
        ]
      }
    }
  }
}
```

Before the change
```
{
  "data": {
    "package": {
      "modules": {
        "nodes": [
          {
            "name": "b",
            "package": {
              "address": "0x1f7f5733d11a5556cddcf73d4642351159b77aafe1b7c5d3d0b2e310f315d4e6",
              "linkage": [
                {
                  "originalId": "0x0000000000000000000000000000000000000000000000000000000000000001",
                  "upgradedId": "0x0000000000000000000000000000000000000000000000000000000000000001",
                  "version": 1
                },
                {
                  "originalId": "0x0000000000000000000000000000000000000000000000000000000000000002",
                  "upgradedId": "0x0000000000000000000000000000000000000000000000000000000000000002",
                  "version": 1
                },
                {
                  "originalId": "0x40adad4ef6d528484468adabe558d051a1d98cd9202de1f4ed3d71724f3f43c3",
                  "upgradedId": "0x922ac7f39faf25f030db32835219e7deb7b87b08f1a7703e1b6e5beb6b5a96f2",
                  "version": 2
                }
              ]
            }
          }
        ]
      }
    }
  }
}
```

B's source code is:
```move
module b::b {
    use std::string::{Self, String};
    use a_v1::a_v1;

    public fun hello_default(ctx: &mut TxContext) {
        let name = string::utf8(b"hello");
    }

    //public fun test(ctx: &mut TxContext): a_v1::Object {
    //    a_v1::default(ctx)
    //}
}
```

B's Move.toml dependency list is:
```
[dependencies]
sui = { git = "https://github.com/mystenlabs/sui.git", subdir = "crates/sui-framework/packages/sui-framework", rev = "framework/testnet" }
a_v1 = { local = "../a_v2" }
```

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [x] CLI: By default, the `sui client publish / upgrade` commands will not store anymore in the `linkage table` any dependency that is not referenced in the source code.
- [ ] Rust SDK:

### CODE REVIEW

**Comment by stefan-mysten** [comment]
This PR breaks the following tests. Would have to look into them:
```bash
 TRY 4 FAIL [   0.897s] sui-core move_package_tests::package_digest_changes_with_dep_upgrades_and_in_sync_with_move_package_digest
  TRY 4 FAIL [   2.742s] sui-package-resolver tests::test_function_parameters
  TRY 4 FAIL [   2.010s] sui-package-resolver tests::test_relinking_layout
  TRY 4 FAIL [  23.935s] sui-source-validation tests::linkage_differs
  TRY 4 FAIL [  18.823s] sui-source-validation tests::successful_verification_upgrades
  TRY 4 FAIL [  16.837s] sui-source-validation tests::successful_verification_with_bytecode_dep
```

File: crates/sui-move-build/src/lib.rs [hunk 1]
```diff
@@ -234,10 +237,9 @@ pub fn build_from_resolution_graph(
     resolution_graph: ResolvedGraph,
     run_bytecode_verifier: bool,
     print_diags_to_stderr: bool,
-    chain_id: Option<String>,
+    published_at: Result<ObjectID, PublishedAtError>,
+    dependencies: PackageDependencies,
```
Comment (amnn) [inline crates/sui-move-build/src/lib.rs:241]:
```suggestion
    mut dependency_ids: PackageDependencies,
```

File: crates/sui-move-build/src/lib.rs [hunk 2]
```diff
@@ -301,10 +303,22 @@ pub fn build_from_resolution_graph(
...
+        })
+        .collect();
+    let mut dependencies = dependencies.clone();
+    dependencies.published.retain(|_, id| deps.contains(id));
+
     Ok(CompiledPackage {
         package,
         published_at,
-        dependency_ids,
+        dependency_ids: dependencies,
```
Comment (amnn) [inline crates/sui-move-build/src/lib.rs:321]:
```suggestion
        dependency_ids,
```

File: crates/sui-move-build/src/lib.rs [hunk 3]
```diff
@@ -301,10 +303,22 @@ pub fn build_from_resolution_graph(
...
+    // Filter out packages that are in the manifest but not referenced in the source code.
+    let deps: BTreeSet<_> = package
+        .all_compiled_units()
+        .filter_map(|x| {
+            let address = x.address.into_inner();
+            (address != AccountAddress::ZERO).then_some(address)
+        })
+        .collect();
+    let mut dependencies = dependencies.clone();
+    dependencies.published.retain(|_, id| deps.contains(id));
```
Comment (amnn) [inline crates/sui-move-build/src/lib.rs:316]:
```suggestion
    dependency_ids.published.retain(|_, id| deps.contains(id));
```

File: crates/sui/tests/cli_tests.rs [hunk 4]
```diff
@@ -1648,12 +1648,9 @@ async fn test_package_publish_nonexistent_dependency() -> Result<(), anyhow::Err
     .execute(context)
     .await;
 
-    let err = result.unwrap_err().to_string();
-    assert!(
-        err.contains("Dependency object does not exist or was deleted"),
-        "{}",
-        err
-    );
+    // publishing with a dependency that is not referenced in code is OK, as the dependency list
```
Comment (amnn) [inline crates/sui/tests/cli_tests.rs:1651]: It seems like in this case you need to update the test such that the dependency is actually referenced but does not exist, because that is what we're trying to test here.

**Review by amnn — APPROVED** [review]
The implementation looks good, thanks @stefan-mysten. When it comes to updating the tests, I suspect many of them are now failing because before they were adding a dependency that was not referenced as a convenient short hand as part of the test set-up.

In most if not all cases the right thing to do is probably to introduce a reference to the non-existent dependency in the Move code, not expect that it will be stripped. I made a comment to that effect on the test that was already updated, but watch out for that when cleaning up the other tests too.

File: crates/sui/tests/cli_tests.rs (hunk 4 above)
Comment (stefan-mysten) [reply crates/sui/tests/cli_tests.rs]: Yes, but now the issue will be that source verification will trigger an error that it cannot find the on-chain version of this dependency. FWIW, that Move package has set a random `published-at` field in the manifest.

Comment (amnn) [reply crates/sui/tests/cli_tests.rs]: It's okay that the error wording is different, the point is that we're testing the error case here.

File: crates/sui/tests/data/module_dependency_nonexistent/Move.lock [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# @generated by Move, please check-in and do not edit manually.
```
Comment (mdgeorge4153) [inline crates/sui/tests/data/module_dependency_nonexistent/Move.lock:1]: I'm not clear on whether this should be checked in. In particular, I'm surprised that the IDs are stable across releases. Maybe this should be ignored?

File: crates/sui/tests/cli_tests.rs [hunk 6]
```diff
@@ -1599,6 +1599,25 @@ async fn test_package_publish_nonexistent_dependency() -> Result<(), anyhow::Err
         .await?
         .data;
 
+    // Get chain identifier
+    let chain_id = client.read_api().get_chain_identifier().await?;
```
Comment (mdgeorge4153) [inline crates/sui/tests/cli_tests.rs:1603]: This test might be easier as a shell test. See https://github.com/MystenLabs/sui/blob/mdgeorge/source-verification-deprecation/crates/sui/tests/shell_tests/with_network/source_verification_deprecation/no_flags.sh for an example

Comment (mdgeorge4153) [reply crates/sui/tests/cli_tests.rs:1603]: At any rate, I'm finding the goal of the test a little hard to follow; a comment explaining it would be helpful

**Review by mdgeorge4153 — COMMENTED** [review]
A few comments on the tests; I am having some trouble following the implementation changes so I didn't review them very carefully. Lmk if you want me to take a closer look.

File: crates/sui/tests/cli_tests.rs (hunk 6 above)
Comment (stefan-mysten) [reply crates/sui/tests/cli_tests.rs:1603]: Basically, the test tries to use an object that does not exist. `Move.lock` keeps the address stable because the package does not get republished every time these tests are run. It's "fixed" if you want, we just have the package and its source code locally. 

When this test is run, will try to use the package with address`0x860faae0287777916571fc1ee341807534f468697519a7e13b8f1b9debbc4e6e` and it will fail because this object does not exist on the local network that this test is running against.

Comment (stefan-mysten) [reply crates/sui/tests/cli_tests.rs:1603]: Regarding the shell test, perhaps it's easier but I'd like to avoid just changing one test now. We'd have to overhaul this whole test suite which is out of scope for this PR.

File: crates/sui/tests/data/module_dependency_nonexistent/Move.lock (hunk 5 above)
Comment (stefan-mysten) [reply crates/sui/tests/data/module_dependency_nonexistent/Move.lock:1]: See my comment above: this package does not get published every time the test runs. It's "locked" on purpose so that we can refer to it in specific test cases.

Comment (stefan-mysten) [reply crates/sui/tests/data/module_dependency_nonexistent/Move.lock:1]: We need the lock file because here we have the `original-published-id`; I remember there was a specific reason why it did not get picked up from `Move.toml`, but don't remember exactly why.

**Comment by stefan-mysten** [comment]
Superseeded by #21168
